crates/hcl-edit/src/parser/expr.rs

```diff
@@ -333,11 +334,25 @@ fn for_list_expr<'i, 's>(
     state: &'s RefCell<ExprParseState>,
 ) -> impl Parser<Input<'i>, (), ContextError> + 's {
     move |input: &mut Input<'i>| {
-        (for_intro, decorated(ws, expr, ws), opt(for_cond))
-            .map(|(intro, value_expr, cond)| {
+        (
+            for_intro,
+            decorated(ws, expr_with_state(state), ws),
+            opt(("...", raw_string(ws))),
+            opt(for_cond(state)),
+        )
+            .map(|(intro, value_expr, grouping, cond)| {
                 let mut expr = ForExpr::new(intro, value_expr);
                 expr.cond = cond;
 
+                if let Some((_, trailing)) = grouping {
+                    expr.grouping = true;
+                    if let Some(ref mut cond) = expr.cond {
+                        cond.decor_mut().set_prefix(trailing);
+                    } else {
+                        expr.decor_mut().set_suffix(trailing);
+                    }
+                }
```

Comment:
this seems... like a terrible approach, because `for_cond` is already supposed to consume leading whitespace, but i don't know how to do the equivalent of a union over two parsers so this is my best guess so far.